sys/usb/usbus/dfu/dfu.c

```diff
@@ -0,0 +1,255 @@
+}
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler)
+{
+    usbus_dfu_device_t *dfu = (usbus_dfu_device_t*) handler;
+
+    /* Set up descriptor generators */
+    dfu->dfu_descr.next = NULL;
+    dfu->dfu_descr.funcs = &_dfu_descriptor;
+    dfu->dfu_descr.arg = dfu;
+
+    /* Configure Interface 0 as control interface */
+    dfu->iface.class = USB_DFU_INTERFACE;
+    dfu->iface.subclass = USB_DFU_SUBCLASS_DFU;
+
+    dfu->iface.protocol = ((dfu->mode) ? USB_DFU_PROTOCOL_DFU_MODE :
+                                USB_DFU_PROTOCOL_RUNTIME_MODE);
+
+    dfu->iface.descr_gen = &dfu->dfu_descr;
+    dfu->iface.handler = handler;
+    dfu->iface.alts = NULL;
+
+    /* Add interfaces to the stack */
+    usbus_add_interface(usbus, &dfu->iface);
+    usbus_handler_set_flag(handler, USBUS_HANDLER_FLAG_RESET);
+}
+
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_control_request_state_t state,
+                          usb_setup_t *setup)
+{
+    (void)usbus;
+    (void)state;
+    (void)handler;
+    (void)setup;
+
+    return 1;
+}
+
+
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                             usbdev_ep_t *ep, usbus_event_transfer_t event)
+{
+    (void)event;
+    (void)usbus;
+    (void)handler;
+    (void)ep;
+}
+
+void dfu_control_req(usbus_t *usbus, usbus_control_handler_t *handler)
+{
+    usb_setup_t *pkt = &handler->setup;
+    usbopt_enable_t enable = USBOPT_DISABLE;
+
+    DEBUG("DFU control request:%d\n", pkt->request);
+    switch(pkt->request) {
+        case DFU_DETACH:
+            /* Detach USB bus */
+            usbdev_set(usbus->dev, USBOPT_ATTACH, &enable, sizeof(usbopt_enable_t));
+            /* EXPERIMENTAL: Start a new usbus instance with a single interface */
+            /* TODO: properly integrate this part to riotboot */
+            usbus_init(&usbus2, usbdev_get_ctx(0));
+            usbus_dfu_init(&usbus2, &dfu2, USBUS_DFU_MODE);
+            usbus_create(_stack, USBUS_STACKSIZE, USBUS_PRIO, USBUS_TNAME, &usbus2);
+            break;
+        case DFU_DOWNLOAD:
+            /* Host indicates end of firmware download */
+            if (pkt->length == 0) {
+                /* Set DFU to manifest sync */
+                dfu_state = USBUS_DFU_STATE_DFU_MANIFEST_SYNC;
+                /* Flash pending buffer */
+                if (offset) {
+                    _flashpage();
+                }
+            }
+            else if (dfu_state != USBUS_DFU_STATE_DFU_DL_SYNC) {
+                dfu_state = USBUS_DFU_STATE_DFU_DL_SYNC;
+            }
+            else {
+                /* Retrieve firmware data */
+                size_t len = 0;
+                usbdev_ep_get(handler->out, USBOPT_EP_AVAILABLE,
+                              &len, sizeof(size_t));
+
+                memcpy(&page_mem[handler->received_len + offset - len],
+                       handler->out->buf, len);
```

Comment:
What is the relationship between `handler->received_len` and `len`?

Reply:
handler->received_len is the endpoint size while len is the size of the frame within the endpoint.
If I find an understandable explanation I'll add it :)